central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,623 @@
+		req.dataStartTime = req.ReportConfig.GetVulnReportFilters().GetSinceStartDate()
+	}
+
+	// Change report status to PREPARING
+	err = s.upsertReportStatus(req.Ctx, req.ReportMetadata, storage.ReportStatus_PREPARING)
+	if err != nil {
+		s.logAndUpsertError(errors.Wrap(err, "Error changing report status to PREPARING"), req)
+		return
+	}
+
+	err = s.generateAndSendReportResult(req)
+	if err != nil {
+		s.logAndUpsertError(err, req)
+		return
+	}
+
+	// Change report status to SUCCESS
+	req.ReportMetadata.ReportStatus.CompletedAt = types.TimestampNow()
+	err = s.upsertReportStatus(req.Ctx, req.ReportMetadata, storage.ReportStatus_SUCCESS)
+	if err != nil {
+		s.logAndUpsertError(errors.Wrap(err, "Error changing report status to SUCCESS"), req)
+		return
+	}
+	s.takeReportSnapshot(req)
+}
+
+func (s *scheduler) notifySchedulerForMoreReports() {
+	s.sendNewReports <- true
+}
+
+func (s *scheduler) removeFromRunningReportConfigs(configID string) {
+	s.schedulerLock.Lock()
+	defer s.schedulerLock.Unlock()
+	s.runningReportConfigs.Remove(configID)
+}
+
+func (s *scheduler) UpsertReportSchedule(reportConfig *storage.ReportConfiguration) error {
+	s.cronJobsLock.Lock()
+	defer s.cronJobsLock.Unlock()
+
+	// Remove the old entry if this is an update
+	if oldEntryID, ok := s.reportConfigToEntryIDs[reportConfig.GetId()]; ok {
+		s.cron.Remove(oldEntryID)
+	}
+	if reportConfig.GetSchedule() != nil {
+		cronSpec, err := schedule.ConvertToCronTab(reportConfig.GetSchedule())
+		if err != nil {
+			return err
+		}
+		entryID, err := s.cron.AddFunc(cronSpec, s.reportClosure(reportConfig))
+		if err != nil {
+			return err
+		}
+		s.reportConfigToEntryIDs[reportConfig.GetId()] = entryID
+	}
+	return nil
+}
+
+func (s *scheduler) RemoveReportSchedule(reportConfigID string) {
+	s.cronJobsLock.Lock()
+	defer s.cronJobsLock.Unlock()
+
+	oldEntryID, exists := s.reportConfigToEntryIDs[reportConfigID]
+	if exists {
+		s.cron.Remove(oldEntryID)
+		delete(s.reportConfigToEntryIDs, reportConfigID)
+	}
+}
+
+/* Functions to add/remove report jobs from queue */
+
+// CancelReport cancels a report that is still waiting in queue.
+// If the report is already being prepared or has completed execution, it cannot be cancelled.
+func (s *scheduler) CancelReport(ctx context.Context, reportID string) error {
+	metadata, found, err := s.reportMetadataStore.Get(ctx, reportID)
+	if err != nil {
+		return errors.Errorf("Error finding report ID '%s': %s", reportID, err)
+	}
+	if !found {
+		return errors.Errorf("Report ID '%s' not found", reportID)
+	}
+	if metadata.ReportStatus.RunState == storage.ReportStatus_SUCCESS ||
+		metadata.ReportStatus.RunState == storage.ReportStatus_FAILURE {
+		return errors.Errorf("Cannot cancel. Report ID '%s' has already been executed", reportID)
+	}
+
+	if !s.tryRemoveReportFromQueue(reportID) {
+		return errors.Errorf("Cannot cancel. Report ID '%s' is already being prepared", reportID)
```

Comment:
If it is already complete, just return

Reply:
This the case when report is being already prepared, and so it cannot be cancelled. In case report is already completed, I changed `return error` to just `return nil`.

Reply:
instead of '%s' use %q which is the same thing